MM-31748 - Add documentation on how to migrate to Mattermost CR (#209)
mattermost/mattermost-operator: docs/migration.md

# Migrating ClusterInstallation to Mattermost

> As of the new release, the Custom Resource managed by the Mattermsot Operator changes from `ClusterInstallation` to `Mattermost`.

Reviewer: ```suggestion
As of the new release, the Custom Resource managed by the Mattermost Operator changes from `ClusterInstallation` to `Mattermost`.
```

Besides the name change, some new functionality is introduced while other is removed.

`BlueGreen` and `Canary` deployments were not widely used and introduced a lot of complexity, so those features were dropped. In most cases, the multi-replica Mattermost cluster proved to be enough for save updates between versions.

> Behavior similar to `BlueGreen` could be mimicked by using multiple Mattermost Instances.

Reviewer: ```suggestion
Behavior similar to `BlueGreen` could be mimicked by using multiple Mattermost resources and with some extra ingresses and services.
```

Add a bit more context

## Automatic migration

> It is possible for the Operator to migrate `ClusterInstallation` to `Mattermost` as long as it does not contain any of the unsupported features (like `BlueGreen` or `Canary`).

Reviewer: ```suggestion
It is possible for the Operator to migrate `ClusterInstallation` to `Mattermost` as long as none of the unsupported features are enabled (like `BlueGreen` or `Canary`).
```

> For migration to be possible, the Mattermost Operator needs to be in version `v1.12.x`.

Reviewer: ```suggestion
For migration to be possible, the Mattermost Operator needs to be version `v1.12.x`.
```

> **NOTE:** Make sure that Mattermost Operator is in version `v1.12.x` before starting the migration.
> To do that, inspect the Operator image tag. To get full image name and tag, run:
> ```
> kubectl -n [OPERATOR_NAMESPACE] get deployment mattermost-operator -o jsonpath='{.spec.template.spec.containers[0].image}'
> ```
> If the Operator's minor version is earlier than `12`, update it to `v1.12.0` by updating manifests [to the released version](https://raw.githubusercontent.com/mattermost/mattermost-operator/v1.12.0/docs/mattermost-operator/mattermost-operator.yaml). 

To run the migration, follow the steps:

1. Prepare the namespace and name of the `ClusterInstallation` and export those as environment variables:
    ```
    export CI_NAMESPACE=[YOUR_NAMESPACE]
    export CI_NAME=[NAME_OF_YOUR_INSTALLATION]
    ```

1. Trigger the migration by setting `spec.migrate` to `true` on the `ClusterInstsallation` instance:
    ```bash
    kubectl -n ${CI_NAMESPACE} patch clusterinstallation ${CI_NAME} --type merge --patch "spec:
      migrate: true"
    ```

> 1. Wait for migration to finish. The migration is done when the `ClusterInstallation` is removed and the `Mattermost` CR with the same name is created and it's `status.state` is equal to `stable`.

Reviewer: Should we mention somewhere that the migration is done without any downtime?

Author: Good point, I have added a sentence.

# Migrating ClusterInstallation to Mattermost

As of the new release, the Custom Resource managed by the Mattermost Operator changes from `ClusterInstallation` to `Mattermost`.

> Besides the name change, some new functionality is introduced while other is removed.

Reviewer 2: ```suggestion
Besides the name change, some new functionality is introduced while other functionality is changed or removed.
```

## Automatic migration
It is possible for the Operator to migrate `ClusterInstallation` to `Mattermost` as long as none of the unsupported features are enabled (like `BlueGreen` or `Canary`). 
During the migration, old Pods are deleted only after the new `Mattermost` resource reaches the `stable` state, 
therefore **the Mattermost instance should not experience any downtime.**

For migration to be possible, the Mattermost Operator needs to be version `v1.12.x`.

> **NOTE:** Make sure that Mattermost Operator is in version `v1.12.x` before starting the migration.
> To do that, inspect the Operator image tag. To get full image name and tag, run:
> ```
> kubectl -n [OPERATOR_NAMESPACE] get deployment mattermost-operator -o jsonpath='{.spec.template.spec.containers[0].image}'
> ```
> If the Operator's minor version is earlier than `12`, update it to `v1.12.0` by updating manifests [to the released version](https://raw.githubusercontent.com/mattermost/mattermost-operator/v1.12.0/docs/mattermost-operator/mattermost-operator.yaml). 

To run the migration, follow the steps:

> 1. Prepare the namespace and name of the `ClusterInstallation` and export those as environment variables:

Reviewer 2: ```suggestion
1. Record the namespace and name of the `ClusterInstallation` and export those as environment variables on the device that is performing the migration:
```

Author: Changed it slightly, as I feel like it could be a little confusing as the actual migration is performed by the Operator, user only triggers it.

Reviewer 2: Oops, yeah that's what I meant to write. Good catch.
